scripts/benchmarks/README.md

```diff
@@ -92,3 +92,102 @@ python benchmark.py --help
+
+After every experiment, 
+  - the logged values are calibrated to remove any existing foreign memory values
+  - the peak values for each gpu device are taken
+  - the values are finally averaged across all devices.
+
+### HuggingFace `HFTrainer`
+HFTrainer has a feature to log memory through the `skip_memory_metrics=False` training argument. In their [documentation](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.TrainingArguments.skip_memory_metrics), it is mentioned that setting this argument to `False` will affect training speed. In our tests so far (below), we do not see significant difference in throughput (tokens/sec) when using this argument.
+
+The HFTrainer API is more granular than `nvidia-smi` as 
+  - It reports the allocated memory by calling `torch.cuda.memory_allocated()` and `torch.cuda.max_memory_allocated()` inside its probes
+  - It has memory logging probes at different stages of the Trainer - `init`, `train`, `evaluate`, `predict` 
+
+##### NOTE:
+When in distributed mode, the Trainer will only log the rank 0 memory.
+
+#### Deciphering the Memory Metrics
+When the memory metrics are logged, additional metrics are included in the output of `Trainer.train / Trainer.evaluate / Trainer.predict`. 
+
+The additional metrics consist of:
+A. absolute memory before trainer initialization
+B. delta of allocated memory of a stage
+C. peaked_delta of a stage (extra mem consumed and freed in between) 
+
+B & C are taken before/after each of the following stages:
+- <init>: `HFTrainer.__init__`
+- <train>: `HFTrainer.train` (only if called)
+- <evaluate>: `HFTrainer.evaluate` (only if called)
+- <predict>: `HFTrainer.predict` (only if called)
+
+Example Output:
+
+```
+output_metrics = {
+    'train_runtime': 143.6704, 
+    'train_samples_per_second': 0.835, 
+    'train_steps_per_second': 0.209, 
+    'train_tokens_per_second': 3421.164, 
+    'train_loss': 1.292790667215983, 
+    'init_mem_cpu_alloc_delta': 8192, 
+    'init_mem_gpu_alloc_delta': 0, 
+    'init_mem_cpu_peaked_delta': 0, 
+    'init_mem_gpu_peaked_delta': 0, 
+    'train_mem_cpu_alloc_delta': 531410944, 
+    'train_mem_gpu_alloc_delta': 126487040, 
+    'train_mem_cpu_peaked_delta': 0, 
+    'train_mem_gpu_peaked_delta': 11016635392, 
+    'before_init_mem_cpu': 5831098368, 
+    'before_init_mem_gpu': 4747206656, 
+    'epoch': 0.04
+}
+```
+
+To compute the total GPU memory allocated for training
+- Total memory delta for train stage = 126487040 bytes ('train_mem_gpu_alloc_delta') + 11016635392 bytes ('train_mem_gpu_peaked_delta')
```

Comment:
we need a link from HF docs to verify that this is indeed the way to intepret these numbers.